File: dynamic_programming/MaxProfit2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        currentHold, currentNotHold = -sys.maxsize, 0
        for price in prices:
            previousHold, previousNotHold = currentHold, currentNotHold
            currentHold = max(previousHold, previousNotHold - price)
            currentNotHold = max(currentNotHold, previousHold + price)
        return currentNotHold if prices else 0

File: dynamic_programming/MaxProfit2.cs
public class Solution {
    public int MaxProfit(int[] prices) {
        // Since it is impossible to sell stock on first day, 
        // set -infinity as initial value for currentHold
        int currentHold = Int32.MinValue, currentNotHold = 0;
        foreach (int price in prices){
            int previousHold = currentHold;
            int previousNotHold = currentNotHold;
            // either keep hold, or buy in stock today at stock price
            currentHold = Math.Max(previousHold, previousNotHold - price);
            // either keep not-hold, or sell out stock today at stock price
            currentNotHold = Math.Max(previousNotHold, previousHold + price);
        }
        return prices.Any() ? currentNotHold : 0; 
    }
}